Samples Revamp: ML - Batch Sentiment Analysis (#320)
dotnet/spark: examples/Microsoft.Spark.CSharp.Examples/MachineLearning/Sentiment/README.md

### 4. Add ML.NET to .NET for Apache Spark App

You have a few options to start creating a .NET for Apache Spark app that uses this ML.NET code and trained model. Make sure that in any app you develop, you've downloaded the [Microsoft.ML NuGet Package](https://www.nuget.org/packages/Microsoft.ML).

#### Option 1: Add Projects

One option is to use Model Builder's *Add Projects* feature, which will result in 3 projects:

* Your original app (**myMLApp**)
* A console app that allows you to build/train/test the model (**myMLAppML.ConsoleApp**)
* A .NET Standard class library that contains model input/output and your trained model in a zip file (**myMLAppML.Model**)

You would begin writing your .NET for Apache Spark code (and paste in the code generated from Model Builder, shown in step 3 above) in the original app **myMLApp.**

![Model Builder Result](https://dotnet.microsoft.com/static/images/model-builder-generated-code.png?v=iC-r8k3zpKUwQVoNOH34D903IhXhIb4CsX003484s7c)

#### Option 2: Create a new console app (shown in this repo)

> Rather than working with the projects/files produced by Model Builder's Add Projects, you can create a new, separate C# console app. You just need to copy over your model's zip file to a directory your new console app can access. In this repo, a trained model **MLModel.zip** has already been included for you in the *Resources* folder.

Reviewer: ```suggestion
Rather than working with the projects/files produced by Model Builder's Add Projects, you can create a new, separate C# console app. You just need to copy over your model's zip file to a directory your new console app can access. In this repo, a trained model **MLModel.zip** has already been included for you in the [Resources](./Resources) folder.
```

As we create the logic for our Spark app, we'll paste in the code generated from Model Builder and include some other class definitions.

## .NET for Spark

> Now that we've trained an ML.NET model for sentiment analysis, we can begin writing the .NET for Spark code that will read in our Yelp data, pass each review to our ML.NET model, and predict whether reviews are positive or negative.

Reviewer: Just a general ML question.  Is it okay to mix different types of reviews like this ?  The MLModel trained using the Amazon dataset, which are reviews on the quality of a product.   Yelp reviews would be reviews on food/decor/service.

Author: I believe it should be acceptable in this use case. Both datasets seem general enough that the model can learn what a user says when they're positive vs when they're negative about something (whether it's product quality or food/service). 

i.e. it'll catch that terms like "great" "very good" and "ideal" are positive, while "unacceptable" "waste" and "unacceptable" are negative, and both datasets contain terms like these.

Author: Also, I found these datasets through the [ML.NET sentiment analysis tutorial](https://docs.microsoft.com/en-us/dotnet/machine-learning/tutorials/sentiment-analysis), so I also took that as a sign that it should be fine.

Reviewer: Generally in the ML world if there is a small dataset, then they split it into a training set and a test set.   From the link you gave, they only use the yelp dataset.  The ML library looks it reads the yelp dataset and will createa two subsets, a training set (to create the model) and a test set (to do testing and validations).

Author: Yep I see where they have divided Yelp dataset into training and testing. Do you think trying to integrate something like that would be preferable/not too confusing rather than using the 2 separate datasets as the sample stands now?

Reviewer: I've done a little dabbling with ML, but my only concern is that I'm just unsure how legitimate this example would seem to an ML expert.

Author: Yeah I understand what you're saying. I feel like with sentiment analysis it could be okay since we're just focusing on simple text, but to improve accuracy/legitimacy, I'm going to divide the Yelp dataset in half and train with one half of it and test with the other half.

Author: Just updated, thanks for the feedback

### 1. Create a Spark Session

In any Spark application, we need to establish a new SparkSession, which is the entry point to programming Spark with the Dataset and 
DataFrame API.

```CSharp
SparkSession spark = SparkSession

>        .Builder()

Reviewer: ```suggestion
    .Builder()
```
fix spacing with below code as well.

       .GetOrCreate();
```

### 2. Read Input File into a DataFrame

We trained our model with the Amazon data, so let's test how well the model performs by testing it with the Yelp dataset. 

```CSharp
DataFrame df = spark.Read().Csv(<Path to yelp data set>);
```

If we want to specify some other aspects of our data, such as whether it has a header and how we want to deal with its schema, we can set some other options when reading in our data:

```CSharp
DataFrame df = spark
    .Read()
    .Option("header", true)
    .Option("inferSchema", true)
    .Csv(<Path to yelp data set>);
```

### 3. Use a UDF to Access ML.NET

We create a User Defined Function (UDF) that calls the *Sentiment* method on each Yelp review.

```CSharp
spark.Udf().Register<string, bool>("MLudf", (text) => Sentiment(text));
```

*Sentiment* is where we'll call our ML.NET code that was generated from the final step of Model Builder.

```CSharp
MLContext mlContext = new MLContext();
ITransformer mlModel = mlContext.Model.Load(@"Resources\MLModel.zip", out var modelInputSchema);
var predEngine = mlContext.Model.CreatePredictionEngine<Review, ReviewPrediction>(mlModel);

> ```

Reviewer: Slightly different than the code, but I assume it's to make it easier to understand ?

Author: I think it was initially to make it easier, but it just seems like now it'd be easier to include and explain the actual code. I've updated it to reflect the current code in `Program.cs`

You may notice the use of *Review* and *ReviewPrediction.* These are classes we define in our project to represent the review data we're evaluating. 

```CSharp
public class Review
{
      // Represents the input review text
      [LoadColumn(0)]
      public string Column1;
}
```

```CSharp
public class ReviewPrediction : Review
{

> 

Reviewer: ```suggestion
```

       [ColumnName("PredictedLabel")]
       public bool Prediction { get; set; }

       public float Probability { get; set; }

       public float Score { get; set; }
} 
```

### 4. Spark SQL and Running Your Code

Now that you've read in your data and incorporated ML, use Spark SQL to call the UDF that will run sentiment analysis on each row of your DataFrame:

```CSharp

> DataFrame sqlDf = spark.Sql("SELECT _c0, MLudf(_c0) FROM Reviews");

Reviewer: This is different than what's in the code.

Author: Updated it now.

# Sentiment Analysis with Big Data

In this sample, you'll see how to use [.NET for Apache Spark](https://dotnet.microsoft.com/apps/data/spark) 
and [ML.NET](https://dotnet.microsoft.com/apps/machinelearning-ai/ml-dotnet) to determine if 
statements are positive or negative, a task known as **sentiment analysis**.

## Problem

Our goal here is to determine if online reviews are positive or negative. We'll be using .NET for Apache Spark to read in a dataset of reviews and ML.NET to perform **binary classification** since categorizing reviews involves choosing one of two groups: positive or negative. You can read more about the problem through the [ML.NET documentation](https://docs.microsoft.com/en-us/dotnet/machine-learning/tutorials/sentiment-analysis).

## Dataset

We'll be using a set of Amazon reviews to train our model and a set of Yelp reviews for testing in our Spark + ML app. You can [download the original data](https://archive.ics.uci.edu/ml/machine-learning-databases/00331/sentiment%20labelled%20sentences.zip) from the [UCI Sentiment Labeled Sentences Dataset]( https://archive.ics.uci.edu/ml/datasets/Sentiment+Labelled+Sentences).

> For the specific ML training/predictions in this app (i.e. when using Model Builder), it helps to have a header for the data. Versions of the Amazon and Yelp datasets with headers can be found in the [Resources](./Resources) folder.

Reviewer: I looked at the `.csv` files with the headers.   Would it be better to have more meaningful names for the columns besides `Column1` `Column2` ?

Author: I agree, I had used those column names while originally writing/testing the app, and now I'll adapt to have better names. Switching column names to `ReviewText` and `Sentiment`.

Follow the [Model Builder Getting Started Guide](https://dotnet.microsoft.com/learn/machinelearning-ai/ml-dotnet-get-started-tutorial/intro) to train your model using the sentiment analysis scenario. 

Follow the steps to:

* Create a new C# Console App
* Pick the **Sentiment Analysis** scenario
* Train using the **Amazon.csv** dataset

![Sentiment Analysis Model Builder](https://dotnet.microsoft.com/static/images/model-builder-vs.png?v=9On8qwmGIXdAyX_-zAmATwYU7fd7tzem-_ojnv1G7XI)

### 3. Generate Model and Code

In the last step of using Model Builder, you'll produce a zip file containing the trained ML.NET model. In the image below, you can see it's contained in **MLModel.zip.**

![ML.NET Zip and Files](https://github.com/bamurtaugh/spark/blob/SparkMLNet/examples/Microsoft.Spark.CSharp.Examples/MachineLearning/images/modelbuilder5proj.PNG)

You'll also generate C# code you can use to consume your model in other .NET apps, like the Spark app we'll be creating.

![Generated Code](https://github.com/bamurtaugh/spark/blob/SparkMLNet/examples/Microsoft.Spark.CSharp.Examples/MachineLearning/images/modelbuilder5code.PNG)

### 4. Add ML.NET to .NET for Apache Spark App

> You have a few options to start creating a .NET for Apache Spark app that uses this ML.NET code and trained model. Make sure that in any app you develop, you've downloaded the [Microsoft.ML NuGet Package](https://www.nuget.org/packages/Microsoft.ML).

Reviewer: ```suggestion
You have a few options to start creating a .NET for Apache Spark app that uses this ML.NET code and trained model. Make sure that in any app you develop, you've added the [Microsoft.ML.FastTree NuGet Package](https://www.nuget.org/packages/Microsoft.ML.FastTree).
```

Reviewer: Do you know why we need FastTree?   Does that mean the `Next Steps` in the screenshot in section 3 is incorrect ?

Author: So I just retrained the ML model with Model Builder (since I needed to train it with the new column names), and the recommended algorithm changed. I'm thinking that depending upon which algorithm is chosen and thus used within the trained model, that affects which nugets the user needs to download.

I referenced the new MLModel.zip I generated, and tried removing the reference to `Microsoft.ML.FastTree` and just including `Microsoft.ML` instead, and it worked. So I think it depends on the model.

I'll add an explanation about this to the readme.

Author: Added an explanation, let me know what you think.

Reviewer: Sounds good!

Author: Great, thanks! Marking this as resolved then (but let me know if you have any other feedback about it).

![ML.NET Zip and Files](https://github.com/bamurtaugh/spark/blob/SparkMLNet/examples/Microsoft.Spark.CSharp.Examples/MachineLearning/images/modelbuilder5proj.PNG)

You'll also generate C# code you can use to consume your model in other .NET apps, like the Spark app we'll be creating.

![Generated Code](https://github.com/bamurtaugh/spark/blob/SparkMLNet/examples/Microsoft.Spark.CSharp.Examples/MachineLearning/images/modelbuilder5code.PNG)

### 4. Add ML.NET to .NET for Apache Spark App

You have a few options to start creating a .NET for Apache Spark app that uses this ML.NET code and trained model. Make sure that in any app you develop, you've added the [Microsoft.ML NuGet Package](https://www.nuget.org/packages/Microsoft.ML).

Depending upon the algorithm Model Builder chooses for your ML model, you may need to add an additional nuget reference in [Microsoft.Spark.CSharp.Examples.csproj](../../Microsoft.Spark.CSharp.Examples.csproj). For instance, if you get an error message that Microsoft.ML.FastTree cannot be found when running your Spark app, you need to add that nuget to your csproj file:

> `<PackageReference Include="Microsoft.ML.FastTree" Version="1.3.1" />`

Reviewer 2: this seems different from what we are referencing in our csproj?

Author: Steve and I learned that depending on what algorithm is used in the model, you may have to add a different reference in your .csproj. Originally we needed `FastTree`, but once I changed to a new model that used a different algorithm, it no longer used `FastTree` and thus we could just reference `Microsoft.ML`. In case users train and use their own model and then run into issues, we wanted to add this explanation.

```CSharp
DataFrame df = spark
    .Read()
    .Option("header", true)
    .Option("inferSchema", true)
    .Csv(<Path to yelp testing data set>);
```

### 3. Use a UDF to Access ML.NET

We create a User Defined Function (UDF) that calls the *Sentiment* method on each Yelp review.

```CSharp
spark.Udf().Register<string, bool>("MLudf", (text) => Sentiment(text));
```

*Sentiment* is where we'll call our ML.NET code that was generated from the final step of Model Builder. The initial code in *Sentiment* sets up the necessary context for ML.NET to perform its prediction:

```CSharp
MLContext mlContext = new MLContext();

ITransformer mlModel = mlContext
    .Model
    .Load(modelPath, out var modelInputSchema);

PredictionEngine<Review, ReviewPrediction> predEngine = mlContext
    .Model
    .CreatePredictionEngine<Review, ReviewPrediction>(mlModel);
```

You may notice the use of *Review* and *ReviewPrediction.* These are classes we define in our project to represent the review data we're evaluating:

```CSharp
public class Review
{
      // Represents the input review text
      [LoadColumn(0)]

>       public string Column1;

Reviewer: ```suggestion
      public string ReviewText;
```
